model/task_lifecycle.go

```diff
@@ -31,111 +31,123 @@ type StatusChanges struct {
+				if err = DequeueAndRestartForTask(nil, &t, message.GithubStateError, caller, fmt.Sprintf("deactivated by '%s'", caller)); err != nil {
+					catcher.Add(err)
+				}
 			}
-		} else {
 			tasksToActivate = append(tasksToActivate, originalTasks...)
+		} else {
+			return nil
 		}
-		if err := task.ActivateTasks(tasksToActivate, time.Now(), caller); err != nil {
-			return errors.Wrapf(err, "can't activate tasks")
-		}
-
-		if t.DistroId == "" && !t.DisplayOnly {
-			grip.Critical(message.Fields{
-				"message": "task is missing distro id",
-				"task_id": t.Id,
-			})
+		if t.IsPartOfDisplay() {
+			if err := UpdateDisplayTaskForTask(&t); err != nil {
+				catcher.Add(errors.Wrap(err, "problem updating display task"))
+			}
 		}
+	}
+	if catcher.HasErrors() {
+		return catcher.Resolve()
+	}
 
-		version, err := VersionFindOneId(t.Version)
-		if err != nil {
-			return errors.Wrapf(err, "error find associated version")
+	if active {
+		if err := task.ActivateTasks(tasksToActivate, time.Now(), caller); err != nil {
+			return errors.Wrapf(err, "can't activate tasks")
 		}
-		if version == nil {
-			return errors.Errorf("could not find associated version : `%s`", t.Version)
+		versionIdsToActivate := []string{}
+		for v := range versionIdsSet {
+			versionIdsToActivate = append(versionIdsToActivate, v)
 		}
-		if err = version.SetActivated(); err != nil {
+		if err := ActivateVersions(versionIdsToActivate); err != nil {
 			return errors.Wrapf(err, "Error marking version as activated")
 		}
-
-		// If the task was not activated by step back, and either the caller is not evergreen
-		// or the task was originally activated by evergreen, deactivate the task
-	} else if !evergreen.IsSystemActivator(caller) || evergreen.IsSystemActivator(t.ActivatedBy) {
-		var err error
-		// deactivate later tasks in the group as well, since they won't succeed without this one
-		if t.IsPartOfSingleHostTaskGroup() {
-			tasksInGroup, err := task.FindTaskGroupFromBuild(t.BuildId, t.TaskGroup)
-			if err != nil {
-				return errors.Wrapf(err, "error finding task group '%s'", t.TaskGroup)
-			}
-			for _, taskInGroup := range tasksInGroup {
-				if taskInGroup.TaskGroupOrder > t.TaskGroupOrder {
-					originalTasks = append(originalTasks, taskInGroup)
-				}
-			}
-		}
-		err = task.DeactivateTasks(originalTasks, caller)
-		if err != nil {
+	} else {
+		if err := task.DeactivateTasks(tasksToActivate, caller); err != nil {
 			return errors.Wrap(err, "error deactivating task")
 		}
-
-		if t.Requester == evergreen.MergeTestRequester {
-			if err = DequeueAndRestartForTask(nil, t, message.GithubStateError, caller, fmt.Sprintf("deactivated by '%s'", caller)); err != nil {
-				return err
-			}
-		}
-	} else {
-		return nil
 	}
 
-	if t.IsPartOfDisplay() {
-		if err := UpdateDisplayTaskForTask(t); err != nil {
-			return errors.Wrap(err, "problem updating display task")
+	for b := range buildIdsSet {
+		t := buildToTaskMap[b]
```

Comment:
Do we need both buildIdsSet and buildToTaskMap? Seems like we could just iterate through buildToTaskMap instead